bakery/checker.go

```diff
@@ -0,0 +1,551 @@
+	}
+	logger.Infof("operations needed after authz macaroons: %#v", need)
+
+	// Try to authorize the operations even even if we haven't got an authenticated user.
+	oks, caveats, err := a.p.Authorizer.Authorize(ctxt, a.identity, need)
+	if err != nil {
+		return authed, used, errgo.Notef(err, "cannot check permissions")
+	}
+	if len(oks) != len(need) {
+		return authed, used, errgo.Newf("unexpected slice length returned from Allow (got %d; want %d)", len(oks), len(need))
+	}
+
+	stillNeed := make([]Op, 0, len(need))
+	for i, ok := range oks {
+		if ok {
+			authed[needIndex[i]] = true
+		} else {
+			stillNeed = append(stillNeed, ops[needIndex[i]])
+		}
+	}
+	if len(stillNeed) == 0 && len(caveats) == 0 {
+		// No more ops need to be authenticated and no caveats to be discharged.
+		return authed, used, nil
+	}
+	logger.Infof("operations still needed after auth check: %#v", stillNeed)
+	if a.identity == nil && len(a.identityCaveats) > 0 {
+		return authed, used, &DischargeRequiredError{
+			Message: "authentication required",
+			Ops:     []Op{LoginOp},
+			Caveats: a.identityCaveats,
+		}
+	}
+	if len(caveats) == 0 {
+		allErrors := make([]error, 0, len(a.initErrors)+len(errors))
+		allErrors = append(allErrors, a.initErrors...)
+		allErrors = append(allErrors, errors...)
+		if len(allErrors) > 0 {
+			// TODO return all errors?
+			logger.Infof("all errors: %q", allErrors)
+			return authed, used, errgo.Notef(allErrors[0], "verification failed")
+		}
+		return authed, used, ErrPermissionDenied
+	}
+	return authed, used, &DischargeRequiredError{
+		Message: "some operations have extra caveats",
+		Ops:     ops,
+		Caveats: caveats,
+	}
+}
+
+// AllowCapability checks that the user is allowed to perform all the
+// given operations. If not, the error will be as returned from Allow.
+//
+// If AllowCapability succeeds, it returns a list of first party caveat
+// conditions that must be applied to any macaroon granting capability
+// to execute the operations.
+//
+// If ops contains LoginOp, the user must have been authenticated with a
+// macaroon associated with the single operation LoginOp only.
+func (a *AuthChecker) AllowCapability(ctxt context.Context, ops ...Op) ([]string, error) {
+	nops := 0
+	for _, op := range ops {
+		if op != LoginOp {
+			nops++
+		}
+	}
+	if nops == 0 {
+		return nil, errgo.Newf("no non-login operations required in capability")
+	}
+	_, used, err := a.allowAny(ctxt, ops)
+	if err != nil {
+		logger.Infof("allowAny returned used %v; err %v", used, err)
+		return nil, errgo.Mask(err, isDischargeRequiredError)
+	}
+	var squasher caveatSquasher
+	for i, isUsed := range used {
+		if !isUsed {
+			continue
+		}
+		for _, cond := range a.conditions[i] {
+			squasher.add(cond)
+		}
+	}
+	return squasher.final(), nil
+}
+
+// caveatSquasher rationalizes first party caveats created for a capability
+// by:
+//	- including only the earliest time-before caveat.
+//	- excluding allow and deny caveats (operations are checked by
```

Comment:
This worries me slightly. This is the only place in bakery itself where the contents of caveat strings are understood. Perhaps this stuff should be in the checkers package somehow.